src/bench/ccoins_caching.cpp

```diff
@@ -6,82 +6,216 @@
-// many times micro-benchmarks of the database showed completely different
-// characteristics than e.g. reindex timings. But that's not a requirement of
-// every benchmark."
-// (https://github.com/bitcoin/bitcoin/issues/7883#issuecomment-224807484)
-static void CCoinsCaching(benchmark::State& state)
+static std::vector<COutPoint>
+SetupDummyCoins(CBasicKeyStore& keystoreRet, CCoinsViewCache& coinsViewCache, int n_coins, int n_keys)
 {
+    std::vector<COutPoint> outpoints;
+    outpoints.resize(n_coins);
+    
+    const std::vector<CKey> keys = SetupDummyKeys(keystoreRet, n_keys);
+    
+    const std::vector<CTransaction> transactions = SetupDummyTransactions(keys, n_coins);
+        
+    for (const CTransaction tx : transactions) {
+        AddCoins(coinsViewCache, tx, 0);
+        outpoints.push_back(COutPoint(tx.GetHash(), 0));
+    }
+    
+    return outpoints;
+}
+
+// Add coins to cache that doesn't exist on disk.
+static void CCoinsViewCacheAddCoinFresh(benchmark::State& state)
+{
+    int n_keys = 1000;
+    int n_txs = N_DEFAULT_CACHE_ENTRIES * N_CACHE_SCALE;
+    
+    // Ignore scaling parameter.
+    state.m_scaling = 1;
+    state.m_num_iters = n_txs;
+    state.m_num_iters_left = n_txs;
+
+    CBasicKeyStore keystore;
+    
+    const std::vector<CKey> keys = SetupDummyKeys(keystore, n_keys);
+    
+    const std::vector<CTransaction> transactions = SetupDummyTransactions(keys, n_txs);
+
+    while (state.IsNewEval()) {
+        CCoinsView coinsView;
+        CCoinsViewCache coinsViewCache(&coinsView);
+        
+        int i = 0;
+        
+        state.ResetTimer();
+                
+        // Benchmark:
+        while (state.KeepRunning()) {
+            AddCoins(coinsViewCache, transactions[i], 0);
+            
+            if (state.IsLastIteration()) {
+                // fprintf(stderr, "Cached coins: %u\n", coinsViewCache.GetCacheSize());
+                // fprintf(stderr, "Cache size: %zu MiB\n", coinsViewCache.DynamicMemoryUsage() / 1024 / 1024);
+                break;
+            }
+            
+            i++;
+        }
+    }
+}
+
+// Flush cache
+static void CCoinsViewCacheFlush(benchmark::State& state)
+{
+    int n_keys = 1000;
+    int n_txs = N_DEFAULT_CACHE_ENTRIES * N_CACHE_SCALE;
+    
+    // Ignore scaling parameter.
+    state.m_scaling = 1;
+    state.m_num_iters = n_txs;
+    state.m_num_iters_left = n_txs;
+
+    CBasicKeyStore keystore;
+    
+    const std::vector<CKey> keys = SetupDummyKeys(keystore, n_keys);
+    
+    const std::vector<CTransaction> transactions = SetupDummyTransactions(keys, n_txs);
+    
+
+    while (state.IsNewEval()) {
+        // TODO: dedup from test suite
+        SelectParams(CBaseChainParams::REGTEST);
+        // const CChainParams& chainparams = Params();
+        ClearDatadirCache();
+        fs::path pathTemp = fs::temp_directory_path() / "bench_bitcoin" / strprintf("%lu_%i", (unsigned long)GetTime(), (int)(InsecureRandRange(1 << 30)));
```

Comment:
```
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/ccoins_caching.cpp:149:  Use int16/int64/etc, rather than the C type long  [runtime/int] [4]
```